File: crates/bookclerk-plugin-sdk/src/tools/package.rs
//! `bookclerk-plugin package`

use std::fs::File;
use std::io::Read;
use std::path::{Path, PathBuf};

use bookclerk_plugin_manifest::{parse, PluginRuntimeKind};
use flate2::write::GzEncoder;
use flate2::Compression;
use sha2::{Digest, Sha256};

use crate::error::{Result, SdkError};

/// Package a plugin directory into `out_dir`, returning the archive path.
pub fn package_plugin(plugin_dir: &Path, out_dir: &Path) -> Result<PathBuf> {
    let toml_path = plugin_dir.join("plugin.toml");
    let text = std::fs::read_to_string(&toml_path)
        .map_err(|e| SdkError::message(format!("read {}: {e}", toml_path.display())))?;
    let manifest = parse(&text).map_err(|e| SdkError::message(e.to_string()))?;
    let version = manifest.version.clone().unwrap_or_else(|| "0.0.0".into());
    let id = &manifest.id;
    std::fs::create_dir_all(out_dir).map_err(SdkError::from)?;

    let staging = out_dir.join(format!(".staging-{id}"));
    let _ = std::fs::remove_dir_all(&staging);
    std::fs::create_dir_all(&staging).map_err(SdkError::from)?;
    std::fs::copy(&toml_path, staging.join("plugin.toml")).map_err(SdkError::from)?;

    if let Some(logo) = manifest.logo.as_deref() {
        if let bookclerk_plugin_manifest::LogoKind::EmbeddedPath(rel) =
            bookclerk_plugin_manifest::validate_logo(logo)
                .map_err(|e| SdkError::message(e.to_string()))?
        {
            let src = plugin_dir.join(&rel);
            if !src.is_file() {
                return Err(SdkError::message(format!(
                    "embedded logo missing for package: {}",
                    src.display()
                )));
            }
            let dest = staging.join(&rel);
            if let Some(parent) = dest.parent() {
                std::fs::create_dir_all(parent).map_err(SdkError::from)?;
            }
            std::fs::copy(&src, &dest).map_err(SdkError::from)?;
        }
    }

    let archive_stem = match manifest.runtime {
        PluginRuntimeKind::Native => {
            let cmd = manifest
                .command
                .as_ref()
                .ok_or_else(|| SdkError::message("native plugin missing command"))?;
            let src = if cmd.is_absolute() {
                cmd.clone()
            } else {
                plugin_dir.join(cmd)
            };
            if !src.is_file() {
                return Err(SdkError::message(format!(
                    "native binary not found for package: {}",
                    src.display()
                )));
            }
            let bin_name = src
                .file_name()
                .ok_or_else(|| SdkError::message("binary name"))?
                .to_os_string();
            let dest = staging.join(&bin_name);
            std::fs::copy(&src, &dest).map_err(SdkError::from)?;
            #[cfg(unix)]
            {
                use std::os::unix::fs::PermissionsExt;
                let mut perms = std::fs::metadata(&dest)?.permissions();
                perms.set_mode(0o755);
                std::fs::set_permissions(&dest, perms)?;
            }
            let triple = host_bookclerk_target();
            format!("bookclerk-plugin-{id}-{version}-{triple}")
        }
        PluginRuntimeKind::Workerd => {
            let w = manifest
                .workerd
                .as_ref()
                .ok_or_else(|| SdkError::message("workerd config missing"))?;
            let modules_src = plugin_dir.join(&w.modules_dir);
            let modules_dst = staging.join(&w.modules_dir);
            copy_dir_recursive(&modules_src, &modules_dst)?;
            // BookclerkPlugin is imported from `@bookclerk/plugin-sdk/workerd`;
            // `bookclerk-workerd` injects that module at runtime.
            format!("bookclerk-plugin-{id}-{version}-workerd")
        }
    };

    let archive_name = format!("{archive_stem}.tar.gz");
    let archive_path = out_dir.join(&archive_name);
    write_tar_gz(&staging, &archive_path)?;
    let _ = std::fs::remove_dir_all(&staging);

    let sums = out_dir.join("SHA256SUMS");
    let digest = sha256_file(&archive_path)?;
    let line = format!("{digest}  {archive_name}\n");
    // Append or replace single-line sums for this archive.
    let mut body = String::new();
    if sums.is_file() {
        body = std::fs::read_to_string(&sums).unwrap_or_default();
        body = body
            .lines()
            .filter(|l| !l.ends_with(&archive_name))
            .collect::<Vec<_>>()
            .join("\n");
        if !body.is_empty() && !body.ends_with('\n') {
            body.push('\n');
        }
    }
    body.push_str(&line);
    std::fs::write(&sums, body).map_err(SdkError::from)?;

    Ok(archive_path)
}

fn host_bookclerk_target() -> String {
    let os = std::env::consts::OS;
    let arch = std::env::consts::ARCH;
    match (os, arch) {
        ("linux", "x86_64") => "linux-x64-gnu".into(),
        ("linux", "aarch64") => "linux-arm64".into(),
        ("macos", "aarch64") => "macos-arm64".into(),
        ("macos", "x86_64") => "macos-x64".into(),
        ("windows", "x86_64") => "windows-x64".into(),
        _ => format!("{os}-{arch}"),
    }
}

fn copy_dir_recursive(src: &Path, dst: &Path) -> Result<()> {
    std::fs::create_dir_all(dst).map_err(SdkError::from)?;
    for entry in std::fs::read_dir(src).map_err(SdkError::from)? {
        let entry = entry.map_err(SdkError::from)?;
        let ty = entry.file_type().map_err(SdkError::from)?;
        let to = dst.join(entry.file_name());
        if ty.is_dir() {
            copy_dir_recursive(&entry.path(), &to)?;
        } else {
            std::fs::copy(entry.path(), to).map_err(SdkError::from)?;
        }
    }
    Ok(())
}

fn write_tar_gz(staging: &Path, archive_path: &Path) -> Result<()> {
    let file = File::create(archive_path).map_err(SdkError::from)?;
    let enc = GzEncoder::new(file, Compression::default());
    let mut builder = tar::Builder::new(enc);
    builder
        .append_dir_all(".", staging)
        .map_err(|e| SdkError::message(format!("tar: {e}")))?;
    let enc = builder
        .into_inner()
        .map_err(|e| SdkError::message(format!("tar finish: {e}")))?;
    enc.finish()
        .map_err(|e| SdkError::message(format!("gzip finish: {e}")))?;
    Ok(())
}

fn sha256_file(path: &Path) -> Result<String> {
    let mut f = File::open(path).map_err(SdkError::from)?;
    let mut hasher = Sha256::new();
    let mut buf = [0u8; 8192];
    loop {
        let n = f.read(&mut buf).map_err(SdkError::from)?;
        if n == 0 {
            break;
        }
        hasher.update(&buf[..n]);
    }
    Ok(hex::encode(hasher.finalize()))
}

File: packages/plugin-sdk/src/tools/package.ts
/**
 * Packages a plugin directory into a distributable `.tar.gz` archive.
 */

import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { spawnSync } from "node:child_process";
import { parse as parseToml } from "smol-toml";
import { validateLogo, validateManifest, type Manifest } from "./validate.js";
import { assertPathInside, refuseSymlinkPath } from "../sparse-workerd/ensure.js";

function hostTarget(): string {
  const plat = process.platform;
  const arch = process.arch;
  if (plat === "linux" && arch === "x64") return "linux-x64-gnu";
  if (plat === "linux" && arch === "arm64") return "linux-arm64";
  if (plat === "darwin" && arch === "arm64") return "macos-arm64";
  if (plat === "darwin" && arch === "x64") return "macos-x64";
  if (plat === "win32" && arch === "x64") return "windows-x64";
  return `${plat}-${arch}`;
}

/**
 * Copy a directory tree, refusing symlinks and non-file/non-dir entries.
 *
 * @param src - Source directory (must not itself be a symlink).
 * @param dst - Destination directory to create.
 * @throws {Error} When a symlink or unsupported type is found.
 */
function copyRecursiveNoSymlinks(src: string, dst: string): void {
  const srcSt = fs.lstatSync(src);
  if (srcSt.isSymbolicLink()) {
    throw new Error(`refusing symlink package source: ${src}`);
  }
  if (!srcSt.isDirectory()) {
    throw new Error(`package source is not a directory: ${src}`);
  }
  fs.mkdirSync(dst, { recursive: true });
  for (const ent of fs.readdirSync(src, { withFileTypes: true })) {
    const from = path.join(src, ent.name);
    const to = path.join(dst, ent.name);
    if (ent.isSymbolicLink()) {
      throw new Error(`refusing symlink in package source: ${from}`);
    }
    if (ent.isDirectory()) {
      copyRecursiveNoSymlinks(from, to);
    } else if (ent.isFile()) {
      fs.copyFileSync(from, to);
    } else {
      throw new Error(`refusing unsupported package source type: ${from}`);
    }
  }
}

/**
 * Packages `pluginDir` into `outDir` and returns the archive path.
 *
 * Native archives include the command binary and are tagged with the host
 * target triple. Workerd archives include the modules tree (the SDK is injected
 * by `bookclerk-workerd` at serve time). Updates `SHA256SUMS` beside the
 * archive.
 *
 * Absolute native `command` paths are treated as operator-selected build
 * outputs. Relative package sources under the plugin tree refuse symlinks so
 * outside bytes cannot enter the archive.
 *
 * @param pluginDir - Plugin root containing `plugin.toml`.
 * @param outDir - Destination directory for the archive and checksums.
 * @returns Absolute path to the created `.tar.gz`.
 * @throws {Error} When validation fails, assets are missing, or `tar` fails.
 *
 * @example
 * ```ts
 * const archive = packagePlugin("./my-plugin", "./dist");
 * console.log(`packed ${archive}`);
 * ```
 */
export function packagePlugin(pluginDir: string, outDir: string): string {
  // Operator-selected root: resolve once so a symlinked plugin directory works.
  const root = fs.realpathSync(path.resolve(pluginDir));
  const tomlPath = assertPathInside(root, "plugin.toml");
  const m = parseToml(fs.readFileSync(tomlPath, "utf8")) as Manifest;
  validateManifest(m);
  const version = m.version ?? "0.0.0";
  const id = m.id;
  fs.mkdirSync(outDir, { recursive: true });
  const staging = path.join(outDir, `.staging-${id}`);
  fs.rmSync(staging, { recursive: true, force: true });
  fs.mkdirSync(staging, { recursive: true });
  fs.copyFileSync(tomlPath, path.join(staging, "plugin.toml"));

  if (m.logo != null) {
    const logo = validateLogo(String(m.logo));
    if (logo.kind === "embedded") {
      const src = assertPathInside(root, logo.value);
      refuseSymlinkPath(root, src);
      if (!fs.existsSync(src) || fs.lstatSync(src).isSymbolicLink() || !fs.statSync(src).isFile()) {
        throw new Error(`embedded logo missing for package: ${src}`);
      }
      const dest = assertPathInside(staging, logo.value);
      fs.mkdirSync(path.dirname(dest), { recursive: true });
      fs.copyFileSync(src, dest);
    }
  }

  const runtime = m.runtime ?? "native";
  let archiveStem: string;
  if (runtime === "native") {
    const cmd = m.command!;
    const absolute = path.isAbsolute(cmd);
    // Absolute command paths are intentional operator-selected build outputs.
    const src = absolute ? path.resolve(cmd) : assertPathInside(root, cmd);
    if (!absolute) {
      refuseSymlinkPath(root, src);
      if (fs.lstatSync(src).isSymbolicLink()) {
        throw new Error(`refusing symlink native command under plugin: ${src}`);
      }
    }
    if (!fs.existsSync(src) || !fs.statSync(src).isFile()) {
      throw new Error(`native binary not found for package: ${src}`);
    }
    const binName = path.basename(src);
    const dest = assertPathInside(staging, binName);
    fs.copyFileSync(src, dest);
    try {
      fs.chmodSync(dest, 0o755);
    } catch {
      /* windows */
    }
    archiveStem = `bookclerk-plugin-${id}-${version}-${hostTarget()}`;
  } else {
    const modulesDir = m.workerd?.modules_dir ?? "modules";
    const srcModules = assertPathInside(root, modulesDir);
    refuseSymlinkPath(root, srcModules);
    const destModules = assertPathInside(staging, modulesDir);
    copyRecursiveNoSymlinks(srcModules, destModules);
    // Authors import `@bookclerk/plugin-sdk/workerd`; bookclerk-workerd injects it.
    archiveStem = `bookclerk-plugin-${id}-${version}-workerd`;
  }

  const archiveName = `${archiveStem}.tar.gz`;
  // Manifest version is free-form and may contain `/` or `..` segments — contain
  // the final name under outDir before any write/delete.
  const outRoot = fs.realpathSync(path.resolve(outDir));
  const archivePath = assertPathInside(outRoot, archiveName);
  const tmpName = `.packaging-tmp-${id}-${process.pid}-${Date.now()}.tar.gz`;
  const tmpPath = assertPathInside(outRoot, tmpName);
  const tar = spawnSync(
    "tar",
    ["-C", staging, "-czf", tmpPath, "."],
    { encoding: "utf8" },
  );
  if (tar.status !== 0) {
    // Attempt-owned temp only — never delete a pre-existing final archive.
    try {
      fs.rmSync(tmpPath, { force: true });
    } catch {
      /* ignore */
    }
    fs.rmSync(staging, { recursive: true, force: true });
    throw new Error(`tar failed: ${tar.stderr || tar.stdout || "status " + tar.status}`);
  }
  fs.renameSync(tmpPath, archivePath);
  fs.rmSync(staging, { recursive: true, force: true });

  const digest = crypto
    .createHash("sha256")
    .update(fs.readFileSync(archivePath))
    .digest("hex");
  const sumsPath = assertPathInside(outRoot, "SHA256SUMS");
  let body = "";
  if (fs.existsSync(sumsPath)) {
    body = fs
      .readFileSync(sumsPath, "utf8")
      .split("\n")
      .filter((l) => l && !l.endsWith(archiveName))
      .join("\n");
    if (body && !body.endsWith("\n")) body += "\n";
  }
  body += `${digest}  ${archiveName}\n`;
  fs.writeFileSync(sumsPath, body);
  return archivePath;
}
